### PR TITLE
[READY]Arconomy: Sales Taggers, split profits on barcoded, sold items!

### PR DESCRIPTION
**Pull request by ArcaneMusic**
I take it back Arconomy is a sick name, ecobbemy eat your heart out

## About The Pull Request

**HEY THERE PUNKS.** Are you RICH? Complaining about how you're not making enough MONEY? Wish that you wern't being held ransom by cargo every shift so that you could buy that pair of insuls you've always dreamed of? Some would say the best way to solve that would be working harder, taking an extra job, or maybe even mining for coins on lavaland. I rejected those answers because they're BROKEN AND STUPID.
Find out how I made my first million with this one neat trick!
BARCODES.
image
This adds a new feature to cargo, where, wrapped packages can have a barcode attached to it. When a barcoded item is sold through cargo, the profits of the of the sale are split between the original scanner and cargo, with the highest personal profit ratio being a 50% even split. Cargo might not mind selling what you have at 50% if there's a bounty for it, but might just rip the tag off it it's a massive stack of plasma sheets.

This is part of my ongoing war with making economy fun and engaging, as a counterbalance to how expensive things are getting. You still have to make things to actually... sell... to cargo mind you, so it's not just free money, but now you have an incentive as well as a way to collect some profit from cargo selling things for you! ...Without mugging the QM, of course.

Part of this implementation includes a new component called pricetag for items contained within a barcoded item. I get the *distinct* feeling that there are some issues with how I've implemented it, so some oversight would be greatly appreciated.
## Why It's Good For The Game

Selling things to cargo has to be done very manually, and most often than not, cargo won't pay you for things you sell to them by default. Now, cargo has an avenue to be able to encourage the crew, in any department, to bring things to them so that cargo profits aren't guided by passive generation or exclusively through bounties. Granted, cargo is still encouraged to do so, but this would allow for expansions on what cargo can sell for a profit, and diversify how the rest of the crew can make a profit mid-round. Collecting things that sell via exports is currently limited to the few things that cargo really loves, tanks, crates, minerals, and the occasional science/botany seeds. 

Plus, unlike most of my features, this will actually encourage inter-departmental as well as inter-personal cooperation!

## To-Do List:

- [x] Send a message to the account holder when their barcoded item has been sold.
- [x] Make a quick piece of floor trash for ripped apart packages so cargo tomfoolery can be observed.
- [x] Add Sales taggers to protolathe designs for supply and cargo. (Or maybe all of them if anyone has an opinion on that)
- [x] Make sure my component isn't held together by sheer force of will
- [x] Make sure that barcoded crates applies the pricetag component to all items within the crate.
- [x] Also uhh... make sure that opening the crate removes the barcoded component.

## Changelog
:cl:
add: The Nanotrasen Deliveries and Handling League (NDHL) have been noticing a decrease in station export profits, and the crate recycling program just isn't cutting it. New Sales Taggers have been deployed in all station cargo setups.
add: Wrapped items can be zapped with Sales Taggers in order to give the account holder a portion of the profit on a sale.
/:cl:

### CODE REVIEW

**Comment by Shadowflame909** [comment]
Yay for capitalism!

**Comment by AtlasRules** [comment]
finally i can rob the ORM of plasma and have enough more for BEPIS machine as scientist

**Comment by MrDoomBringer** [comment]
You should really set it up so that the linked ID card will display a message if the barcode is destroyed or if the item was sold successfully, to encourage people to use this. Other than that, great addition. 👍

**Comment by Tlaltecuhtli** [comment]
add an examinate to parent signal so you can see the % of taxing on the item

**Comment by trollbreeder** [comment]
Guess we're goonstation now.

**Comment by cacogen** [comment]
How does it work? You find an item that's valuable when sold on the cargo shuttle. You package wrap it, and use a barcode tagger on it (where are these?). You then say to cargo, "I've brought you x, it's in here, I promise" and they put it on the shuttle and when it arrives at Centcom half the value goes into your account. There's nothing stopping them from removing the barcode, other than it reducing the likelihood that you'll source goods for them again. Am I correct?

**Comment by Tlaltecuhtli** [comment]
then you dont fetch crates for cargo

**Comment by Krysonism** [comment]
Most QM's will not rip you off.

**Comment by ATH1909** [comment]
Ain't this basically gonna cut cargo's income rate from bounties in half, since ripping off someone package's barcode as a cargonian would be very rude/get you lynched?

File: code/modules/recycling/sortingmachinery.dm [hunk 1]
```diff
@@ -243,3 +292,87 @@
 		var/n = text2num(href_list["nextTag"])
 		currTag = n
 	openwindow(usr)
+
+/obj/item/sales_tagger
+	name = "sales tagger"
+	desc = "A scanner that lets you tag wrapped items for sale, splitting the profit between you and cargo. Ctrl-Click to clear the registered account."
```
Comment (ATH1909) [inline code/modules/recycling/sortingmachinery.dm]: Should "cargo" be capitalized or not? I forgot if we're supposed to capitalize department names.

**Comment by ArcaneMusic** [comment]
> Ain't this basically gonna cut cargo's income rate from bounties in half, since ripping off someone package's barcode as a cargonian would be very rude/get you lynched?
> 
> Eh, it probably won't do that, simply because only 10% of the crew actually reads the changelogs and will remember how to mark something with a barcode/that you can ever mark something with a barcode in the first place when turning a bounty in to cargo. The cargonians certainly aren't gonna remind them...

Well part 2 (tm) of this change is giving a BUNCH more items base export values so that there's an incentive for crew to drop off things for sale. if things that were going unsold all shift have a decent value, then care will wind up making more money than they were before.

**Comment by bluetshirtguy** [comment]
Honestly acquiring a sales tagger should be easy for any crew outside of cargo, I don't know about every protolathe printing one but what about public tool storage or maybe for a small price in a vendor?

**Comment by trollbreeder** [comment]
https://tgstation13.org/wiki/Map_Merger

**Comment by ArcaneMusic** [comment]
Thanks mapmerge hook, real cool of you to stop existing in the one moment I needed you most.

**Comment by trollbreeder** [comment]
Yeah, mapmerge hook is very unreliable.

File: code/modules/recycling/sortingmachinery.dm (hunk 1 above)
Comment (ArcaneMusic) [reply code/modules/recycling/sortingmachinery.dm]: After close examination, no, cargo is not capitalized.

**Comment by Tlaltecuhtli** [comment]
@ArcaneMusic can you add a shortcut where if you hit the floor you tag any parcel on the tile?

**Comment by Ohagi-Chan** [comment]
That's like what mop and soap do, and it's the only way I ever use those tools.

That said, it's a terrible idea in this context, given that it's impossible for there to be >100 things to clean on a single tile, but it is possible to have a ridiculous number of packaged items.

**Comment by Tlaltecuhtli** [comment]
it can only tag 10 items before running out of paper so its self contained

**Comment by ArcaneMusic** [comment]
Well, max is actually 20, but considering that you'd need to wrap up those 20 items with package wrap beforehand, you'd probably want to port that feature onto package wrap first. That said, I might want to hold off on mass barcoding until the inevitable followup PR.

File: code/modules/cargo/exports.dm [hunk 2]
```diff
@@ -43,11 +47,18 @@ Credit dupes that require a lot of manual work shouldn't be removed, unless they
 		var/sold = FALSE
 		if(QDELETED(thing))
 			continue
+
+		//But first I'm going to check if you're getting the full payout.
+		if(thing.GetComponent(/datum/component/pricetag))
+			profit_ratio = (thing.GetComponent(/datum/component/pricetag)).profit_ratio
```
Comment (ninjanomnom) [inline code/modules/cargo/exports.dm]: You shouldn't be using GetComponent to do things that are part of the functionality needed for that component. Use signals to allow the component to interact with the outside without having code outside the component like this.

File: code/datums/components/pricetag.dm [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/datum/component/pricetag
+	var/datum/bank_account/owner = null
```
Comment (ninjanomnom) [inline code/datums/components/pricetag.dm]:
```suggestion
	var/datum/bank_account/owner
```

File: code/datums/components/pricetag.dm [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+	var/profit_ratio = 1
+
+/datum/component/pricetag/Initialize(_owner,_profit_ratio)
+	if(_owner)
+		owner = _owner
```
Comment (ninjanomnom) [inline code/datums/components/pricetag.dm]: You don't need to check if owner is given or not, just assign it

Comment (ninjanomnom) [inline code/datums/components/pricetag.dm]: You should be checking if the parent is an item and returning `COMPONENT_INCOMPATIBLE` if not

File: code/modules/recycling/sortingmachinery.dm [hunk 5]
```diff
@@ -8,18 +8,24 @@
...
 
 /obj/structure/bigDelivery/interact(mob/user)
 	playsound(src.loc, 'sound/items/poster_ripped.ogg', 50, TRUE)
+	new /obj/effect/decal/cleanable/wrapping(get_turf(user))
 	qdel(src)
 
 /obj/structure/bigDelivery/Destroy()
 	var/turf/T = get_turf(src)
+	for(var/obj/I in src.GetAllContents())
+		SEND_SIGNAL(I, COMSIG_ITEM_UNWRAPPED)
```
Comment (ninjanomnom) [inline code/modules/recycling/sortingmachinery.dm]: Why are you doing this?

File: code/modules/recycling/sortingmachinery.dm [hunk 6]
```diff
@@ -8,18 +8,24 @@
...
 	for(var/atom/movable/AM in contents)
 		AM.forceMove(T)
 	return ..()
 
 /obj/structure/bigDelivery/contents_explosion(severity, target)
+	for(var/obj/I in src.GetAllContents())
+		SEND_SIGNAL(I, COMSIG_ITEM_UNWRAPPED)
```
Comment (ninjanomnom) [inline code/modules/recycling/sortingmachinery.dm]: Same here, it isn't being unwrapped so what's this doing here?

File: code/modules/recycling/sortingmachinery.dm [hunk 7]
```diff
@@ -108,17 +158,21 @@
...
+	var/obj/item/barcode/sticker
 
 /obj/item/smallDelivery/contents_explosion(severity, target)
 	for(var/atom/movable/AM in contents)
 		AM.ex_act()
 
 /obj/item/smallDelivery/attack_self(mob/user)
 	user.temporarilyRemoveItemFromInventory(src, TRUE)
+	for(var/obj/I in src.GetAllContents())
+		SEND_SIGNAL(I, COMSIG_ITEM_UNWRAPPED)
```
Comment (ninjanomnom) [inline code/modules/recycling/sortingmachinery.dm]: You have too many places you're using this signal and it looks like you need a more centralized place to handle unwrapping

File: code/modules/recycling/sortingmachinery.dm [hunk 8]
```diff
@@ -243,3 +342,88 @@
...
+	to_chat(user, "<span class='notice'>You clear the registered account.</span>")
+
+/obj/item/sales_tagger/AltClick(mob/user)
+	. = ..()
+	var/potential_cut = input("How much would you like to payout to the registered card?","Percentage Profit") as num|null
+	if(!potential_cut)
+		percent_cut = 50
+	percent_cut = CLAMP(round(potential_cut, 1), 1, 50)
+	to_chat(user, "<span class='notice'>[percent_cut]% profit will be recieved if a package with a barcode is sold.</span>")
+	return
```
Comment (ninjanomnom) [inline code/modules/recycling/sortingmachinery.dm]:
```suggestion

```

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/recycling/sortingmachinery.dm (hunk 6 above)
Comment (ArcaneMusic) [reply code/modules/recycling/sortingmachinery.dm]: This one is my bad, I was under the impression that "Contents_explosion" meant "All the things leaving the wrapped crate".

**Comment by Tlaltecuhtli** [comment]
> Well, max is actually 20, but considering that you'd need to wrap up those 20 items with package wrap beforehand, you'd probably want to port that feature onto package wrap first. That said, I might want to hold off on mass barcoding until the inevitable followup PR.

well wrapping things moves the item to your tile so you can wrap stuff without needing to alt click the tile, maybe do the same so you can just keep clicking on the pile of wrapped items without having to pixel hunt the ones that arent barcoded

**Comment by ArcaneMusic** [comment]
> > Well, max is actually 20, but considering that you'd need to wrap up those 20 items with package wrap beforehand, you'd probably want to port that feature onto package wrap first. That said, I might want to hold off on mass barcoding until the inevitable followup PR.
> 
> well wrapping things moves the item to your tile so you can wrap stuff without needing to alt click the tile, maybe do the same so you can just keep clicking on the pile of wrapped items without having to pixel hunt the ones that arent barcoded

Also keep in mind that if you just want to mass barcode something, you can just put all the items into a single crate/locker and use a barcode on the locker instead, tagging everything within.

**Comment by gitbirb** [comment]
> > > Well, max is actually 20, but considering that you'd need to wrap up those 20 items with package wrap beforehand, you'd probably want to port that feature onto package wrap first. That said, I might want to hold off on mass barcoding until the inevitable followup PR.
> > 
> > 
> > well wrapping things moves the item to your tile so you can wrap stuff without needing to alt click the tile, maybe do the same so you can just keep clicking on the pile of wrapped items without having to pixel hunt the ones that arent barcoded
> 
> Also keep in mind that if you just want to mass barcode something, you can just put all the items into a single crate/locker and use a barcode on the locker instead, tagging everything within.

why not just make a barcoder machine where you swipe your id, and push a crate through, and it wraps and barcodes it to the id?

**Comment by Mickyan** [comment]
you might want to at least add a do_after to uwrapping things or even move it to alt click cause it's really easy to do accidentally

**Comment by spookydonut** [comment]
@ArcaneMusic mergeconflict

**Comment by kriskog** [comment]
Testmerged on bagil. Seems to conflict with #49274

**Comment by Iamgoofball** [comment]
why, exactly is this conflicting

**Comment by kriskog** [comment]
They cant be testmerged together, I didnt look at the details.

**Comment by ExcessiveUseOfCobblestone** [comment]
cringe

**Comment by Arianya** [comment]
They're both touching tgstation.dme so this is kinda unavoidable. Just have to be testmerged separately.

**Comment by Arianya** [comment]
Testmerged on Terry too to get some more test surface area.

**Comment by ExcessiveUseOfCobblestone** [comment]
fix MC then we'll merge

**Pull request merged**